audits/internal4/README.md

```diff
@@ -0,0 +1,126 @@
+via custom stakingToken
+```
+
+Reentrancy low issue: <br>
+```solidity
+ServiceStakingToken.deposit -> SafeTransferLib.safeTransfer(stakingToken, to, amount); -> ServiceStaking.deposit
+via custom stakingToken
+```
+
+Low problem: <br>
+```solidity
+function checkpoint() public returns ()
+uint256 curServiceId; 
+vs
+// Get the current service Id
+uint256 curServiceId = serviceIds[i];
+Details: https://github.com/crytic/slither/wiki/Detector-Documentation#variable-names-too-similar
+Reusing a same variable name in different scopes.
+```
+
+
+```solidity
+    if (state != 4) {
+        revert WrongServiceState(state, serviceId);
+    }
+It's better to use the original type enum.
+Details: https://github.com/pessimistic-io/slitherin/blob/master/docs/magic_number.md
+```
+
+# Low optimization
+```
+if (size > 0) {
+            for (uint256 i = 0; i < size; ++i) {
+                // Agent Ids must be unique and in ascending order
+                if (_stakingParams.agentIds[i] <= agentId) {
+                    revert WrongAgentId(_stakingParams.agentIds[i]);
+                }
+                agentId = _stakingParams.agentIds[i];
+                agentIds.push(agentId);
+            }
+        }
+if size == 0 then for(i = 0; i < 0; i++) -> no loop
+```
+```
+        // Transfer the service for staking
+        IService(serviceRegistry).safeTransferFrom(msg.sender, address(this), serviceId);
+        Last operation?
+```
+
+### General considerations
+Measuring the number of live transactions through a smart contract has fundamental limitations: <br>
+Ethereum smart contracts only have access to the current state - not to historical states at previous blocks. <br>
+Also, there's currently no EVM opcode (hence no Solidity function) to look up the amount of transactions by an address. <br>
+Therefore, we have to rely on the internal counter "nonce" to measure tps (tx per sec). <br>
+https://github.com/safe-global/safe-contracts/blob/1cfa95710057e33832600e6b9ad5ececca8f7839/contracts/Safe.sol#L167 <br>
+
+In this contract, we assume that the services  are mostly honest and are not trying to tamper with this counter. <br>
+ref: IService(serviceRegistry).safeTransferFrom(msg.sender, address(this), serviceId); <br>
+
+There are two ways to fake it: <br>
+1. Fake multisig. Open question
+```solidity
+        (uint96 stakingDeposit, address multisig, bytes32 hash, uint256 agentThreshold, uint256 maxNumInstances, , uint8 state) =
+            IService(serviceRegistry).mapServices(serviceId);
+sInfo.multisig = multisig;
+sInfo.owner = msg.sender;
+uint256 nonce = IMultisig(multisig).nonce();
+Thus, it is trivial to tweak nonce it now if there is some trivial contract with the method setNonce(uint256);
+
+If you wish, you can fight, for example, using this method
+hash(multisig.code) vs well-know hash
+```
+Is it possible to replace the multisig with some kind of fake one without losing the opportunity to receive rewards? <br>
```

Comment:
Before staking.

Reply:
I see, good point. Added this in registries vulnerabilities list as well!

Reply:
So basically for staking to be safe we first need to upgrade the protocol to ensure all services use Safe multisigs.

Reply:
@DavidMinarsch There's no immediate urgency for an upgrade. Specifically, we can perform a direct check on the staking contract when a service is staked under this initial version. Subsequently, we can consider upgrading the GnosisSafeSameAddressMultisig

Reply:
Yes, the verification in this contract will be enough to close the problem now. 
Then we will need to add this check to old contracts. 
Everything is as @mariapiamo said. @kupermind has already made such a fix in post-audit PR.

Reply:
We will do it in the staking contract, however it would be good to update GnosisSafeSameAddressMultisig contract at some point to check for the multisig proxy code.